crates/mysten-metrics/src/guards.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use prometheus::IntGauge;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// Increments gauge when acquired, decrements when guard drops
+pub struct GaugeGuard<'a>(&'a IntGauge);
+
+impl<'a> GaugeGuard<'a> {
+    pub fn acquire(g: &'a IntGauge) -> Self {
+        g.inc();
+        Self(g)
+    }
+}
+
+impl<'a> Drop for GaugeGuard<'a> {
+    fn drop(&mut self) {
+        self.0.dec();
+    }
+}
+
+pub trait GaugeGuardFutureExt: Future + Sized {
```

Comment:
Add a comment for this as well?

Reply:
Done(added comment to the count_in_flight method)